MDANSE/Src/MDANSE/Framework/Configurators/GroupingLevelConfigurator.py

```diff
@@ -92,49 +142,280 @@ def configure(self, value):
+                    / conc
+                )
+
+                output_data.add(
+                    f"{result_name}_[{grp_i}][{grp_j}]_{post_label}",
+                    data_type,
+                    results.shape,
+                    **kwargs,
+                )
+                output_data[f"{result_name}_[{grp_i}][{grp_j}]_{post_label}"][...] = (
+                    post_func(results)
+                )
+                if scaling_factor:
+                    output_data[
+                        f"{result_name}_[{grp_i}][{grp_j}]_{post_label}"
+                    ].scaling_factor = conc
+        else:
+            raise NotImplementedError("Grouped total for dim > 2 not implemented.")
+
+    def pair_labels(
+        self, intra=False, all_pairs=False
+    ) -> list[tuple[str, tuple[str, str]]]:
+        """Generates pair labels.
+
+        Parameters
+        ----------
+        intra : bool
+            Returns the intra label data if true.
+        all_pairs : bool
+            Returns all pairs of labels e.g. OH and HO.
+
+        Returns
+        -------
+        list[tuple[str, tuple[str, str]]]
+            The labels of the results and the labels of the individual
+            atoms in a tuple.
+        """
+        labels = []
+
+        if self["level"] == "atom":
+            atom_selection = self._configurable[self._dependencies["atom_selection"]]
+            selected_elements = atom_selection["unique_names"]
+            for ele_i, ele_j in self.label_pairs(selected_elements, all_pairs):
+                labels.append((f"{ele_i}{ele_j}", (ele_i, ele_j)))
+            return labels
+
+        if intra:
+            for grp in self["group_names"]:
+                eles = sorted(set(self["group_elements"][grp]))
+                for ele_i, ele_j in self.label_pairs(eles, all_pairs):
+                    pair_label = f"[{grp}]_{ele_i}{ele_j}"
+                    label_i = f"[{grp}]_{ele_i}"
+                    label_j = f"[{grp}]_{ele_j}"
+                    labels.append((pair_label, (label_i, label_j)))
+            return labels
+
+        for grp_i, grp_j in self.label_pairs(self["group_names"], all_pairs):
+            eles_i = sorted(set(self["group_elements"][grp_i]))
+            eles_j = sorted(set(self["group_elements"][grp_j]))
+            if grp_i == grp_j and not all_pairs:
+                pairs = it.combinations_with_replacement(eles_i, 2)
+            else:
+                pairs = it.product(eles_i, eles_j)
+            for ele_i, ele_j in sorted(pairs):
+                pair_label = f"[{grp_i}][{grp_j}]_{ele_i}{ele_j}"
+                label_i = f"[{grp_i}]_{ele_i}"
+                label_j = f"[{grp_j}]_{ele_j}"
+                labels.append((pair_label, (label_i, label_j)))
+        return labels
+
+    def update_pair_results(
+        self,
+        calc_func: Callable[[str, str], Iterator[tuple[str, bool, npt.NDArray]]],
```

Comment:
To be precise, calc_func only needs to return an `Iterable` (can be used in a for-loop, i.e. has `__iter__`) rather than an `Iterator` (supports `next`)
```suggestion
        calc_func: Callable[[str, str], Iterable[tuple[str, bool, npt.NDArray]]],
```